[PiperTTS] Initial contribution

GiviMAD opened the pull request:
This is a <s>working in progress</s> PR for integrating Piper.

<s>It's already finished, but I let by error a println call inside the library used, so it can not be merged. I'm going to let it for next weekend so I can also give it another review.</s>

I noticed a problem with Piper when using the low size models at short phrases, but with the medium model I have tested it works great and very fast without using too much ram.

GiviMAD commented:
Hey @dalgwen, if you have a chance to giving it a try, in case you notice something wrong or improvable.
BR.

dalgwen commented:
Excellent !
I found piper-tts some weeks ago and I immediately thought that we should also have a TTS service for it. Well, it's christmas time ! Thanks.
I tested your implementation on my dev env and it works great 👍 
As Mycroft cease activity, I was worried about mimic-tts. But now I can get rid of my mimic docker container in favor of this much simpler solution. I will push this on my "prod" env right now !

About the code, I have suggestions, all about cache : 
- Could you consider extending AbstractCachedTTSService instead of implementing directly TTSService ? It is a smart LRU cache to store the wav files created on disk. As piper is fast and free to use, it is not so important as it is for cloud services, but it seems nonetheless a good idea and it has also some other advantages (like allowing to play concurrently on several sinks with only one call to the TTS service.) 
- Do you think you can cache the json file reading in getAvailableVoices and getVoice ? These two methods are unfortunately called every time by the framework.

Thanks again

dalgwen commented:
> Does it mean we have to remove the binding? Or does it still work?

mimic 3 is completely offline, and the code is 100% open source and on github. So, it will continue to work and to be available.
I don't know if OpenVoice OS or Neon (kind of "successor" of Mycroft) have plan about it, but AFAIK there is no more development on mimic. So it is, at least, on the downard slope.
Piper has the support of rhasspy and Home Assistant, and many qualities, so it is IMHO the perfect choice now.

dalgwen commented:
> Hey @dalgwen, if you have a chance to giving it a try, in case you notice something wrong or improvable. BR.

Additionnal question : Is the binding supposed (do you want the binding) to work on openHAB 4.0 ?

With a 4.0.3 version, I have this : 
```
2023-11-27 14:16:52.141 [WARN ] [org.apache.felix.fileinstall        ] - Error while starting bundle: file:/usr/share/openhab/addons/org.openhab.voice.pipertts-4.1.0-SNAPSHOT.jar
org.osgi.framework.BundleException: Could not resolve module: org.openhab.voice.pipertts [348]
  Unresolved requirement: Import-Package: com.fasterxml.jackson.databind; version="[2.15.0,3.0.0)"

        at org.eclipse.osgi.container.Module.start(Module.java:463) ~[org.eclipse.osgi-3.18.0.jar:?]
        at org.eclipse.osgi.internal.framework.EquinoxBundle.start(EquinoxBundle.java:445) ~[org.eclipse.osgi-3.18.0.jar:?]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundle(DirectoryWatcher.java:1260) ~[?:?]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundles(DirectoryWatcher.java:1233) ~[?:?]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.doProcess(DirectoryWatcher.java:520) ~[?:?]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.process(DirectoryWatcher.java:365) ~[?:?]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.run(DirectoryWatcher.java:316) ~[?:?]
```

GiviMAD commented:
> Excellent ! I found piper-tts some weeks ago and I immediately thought that we should also have a TTS service for it. Well, it's christmas time ! Thanks. I tested your implementation on my dev env and it works great 👍 As Mycroft cease activity, I was worried about mimic-tts. But now I can get rid of my mimic docker container in favor of this much simpler solution. I will push this on my "prod" env right now !
> 
> About the code, I have suggestions, all about cache :
> 
> * Could you consider extending AbstractCachedTTSService instead of implementing directly TTSService ? It is a smart LRU cache to store the wav files created on disk. As piper is fast and free to use, it is not so important as it is for cloud services, but it seems nonetheless a good idea and it also has other advantages (like allowing to play concurrently on several sinks with only one call to the TTS service.) To do this, you can refer to the commit [28e6da6](https://github.com/openhab/openhab-addons/commit/28e6da6342005512805de641f89de2df78b93423) for MaryTTS.
> * Do you think you can cache the json file reading in getAvailableVoices and getVoice ? These two methods are unfortunately called every time by the framework.
> 
> Thanks again

Thank you for giving it a try. And yes make a lot of sense to apply both suggestions. I found an issue with the option for keeping the voice loaded, seems like when I keep one voice in memory and use another there is a bug in the moment of release the memory of the second one, I think I've made something wrong. I hope to find time these weekend to get the PR ready.

> Additionnal question : Is the service supposed (do you want the service) to work on openHAB 4.0 ?

I think bindings of the branch 4.1.0 are not compatibles with 4.0.x versions, I'm not sure about it, I started having problems testing some bindings and moved to the milestones.

GiviMAD commented:
>  I found an issue with the option for keeping the voice loaded, seems like when I keep one voice in memory and use another there is a bug in the moment of release the memory of the second one, I think I've made something wrong. I hope to find time these weekend to get the PR ready.

I found the problem, I incorrectly understood the piper initialize and terminate methods, I have fixed the implementation.

Either way I have changed the preloadModel operation to keep the last voice used loaded, I think it has a lot more sense implemented that way.

Also I've extended the cache class and it works nicely, and added a simple cache mechanism for avoid the json read, thank you for the comments!

wborn commented on bundles/org.openhab.voice.pipertts/pom.xml:
```diff
@@ -0,0 +1,24 @@
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>4.1.0-SNAPSHOT</version>
```
```suggestion
    <version>4.2.0-SNAPSHOT</version>
```

lsiepel reviewed (changes requested):
Mainly some minor text comments. Also also some Exception handling improvements can be made. Half way the PiperTTSService i stopped checking, as it seems consistent and i think if you refactor it, you will fix them.
Hope this makes any sense.

GiviMAD commented on bundles/org.openhab.voice.pipertts/README.md:
```diff
@@ -0,0 +1,60 @@
+# Piper Text-to-Speech
+
+Piper STT integrates [Piper](https://github.com/rhasspy/piper) the fast, local neural text to speech system that sounds great and is optimized for the Raspberry Pi 4 into OpenHAB using a JNI interface.
```
I have rephrased it into multiple sentences.

GiviMAD commented on bundles/org.openhab.voice.pipertts/src/main/java/org/openhab/voice/pipertts/internal/PiperTTSService.java:
```diff
@@ -0,0 +1,436 @@
+    private PiperConfig getPiperConfig() throws IOException {
+        PiperConfig piperConfig = this.piperConfig;
+        if (piperConfig == null) {
+            throw new IOException("Piper config not available");
```
I have updated the library and now the config is not exposed.

lsiepel reviewed (approved):
LGTM, finished the review, but i'm new to voice add-on's, so i think an extra set of @openhab/add-ons-maintainers eyes would be usefull

GiviMAD commented:
> LGTM, finished the review, but i'm new to voice add-on's, so i think an extra set of @openhab/add-ons-maintainers eyes would be usefull

Thank you for the review!

wborn commented on bundles/org.openhab.voice.pipertts/src/main/java/org/openhab/voice/pipertts/internal/PiperTTSService.java:
```diff
@@ -0,0 +1,417 @@
+        piperVoice = piper.loadVoice(voice.voiceModelPath(), voice.voiceModelConfigPath(), voice.speakerId.orElse(-1L));
+        voiceModel = new VoiceModel(voice, piperVoice, piperVoice.getSampleRate(), new AtomicInteger(1), logger);
+        if (config.preloadModel) {
+            synchronized (this) {
```
It is better to create a private lock object and synchronize on that. You never know what other code is also synchronizing on the instance. This prevents unexpected deadlocks.

GiviMAD replied:
I have changed the synchronized blocks to using a ReentrantLock instance. Is that the correct way?

wborn replied:
Since it only does simple locking/unlocking you can simplify it using:

```java
private Object modelLock = new Object();
```

Then you can just use: 

```java
synchronized (modelLock) {
   ....
}
```

That way there will be less code and locking mistakes. :slightly_smiling_face:

GiviMAD replied:
Done. Thank you for the clarification and the rest of the review!

wborn commented on bundles/org.openhab.voice.pipertts/README.md:
```diff
@@ -0,0 +1,58 @@
+
+You can find the link to the available voices at the [Piper README](https://github.com/rhasspy/piper).
+
+Each voice model is composed of two files a onnx runtime model file with extension '.onnx' and a model config file with extension '.onnx.json'.
```
```suggestion
Each voice model is composed of two files an onnx runtime model file with extension '.onnx' and a model config file with extension '.onnx.json'.
```

wborn reviewed (commented):
Thanks for the PR! I've added few more minor comments below:

wborn commented on bundles/org.openhab.voice.pipertts/src/main/java/org/openhab/voice/pipertts/internal/PiperTTSService.java:
```diff
@@ -0,0 +1,423 @@
+    private PiperTTSConfiguration config = new PiperTTSConfiguration();
+    private @Nullable VoiceModel preloadedModel;
+    private @Nullable PiperJNI piper;
+    private HashMap<String, List<Voice>> cachedVoicesByModel = new HashMap<>();
```
```suggestion
    private Map<String, List<Voice>> cachedVoicesByModel = new HashMap<>();
```
